tests/integration_tests/functional/test_balloon.py

```diff
@@ -0,0 +1,687 @@
+    reason="The balloon device is less effective on arm64 architecture"
+)
+def test_rss_memory_lower(test_microvm_with_ssh_and_balloon, network_config):
+    """Check inflating the balloon makes guest use less rss memory."""
+    test_microvm = test_microvm_with_ssh_and_balloon
+    test_microvm.spawn()
+    test_microvm.basic_config()
+    _tap, _, _ = test_microvm.ssh_network_config(network_config, '1')
+    test_microvm.ssh_config['ssh_key_path'] = os.path.join(
+        test_microvm.fsfiles,
+        'debian.rootfs.id_rsa'
+    )
+
+    # Add a memory balloon.
+    response = test_microvm.balloon.put(
+        num_pages=0,
+        deflate_on_oom=True,
+        must_tell_host=False,
+        stats_polling_interval_s=0
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    # Start the microvm.
+    test_microvm.start()
+
+    # Get the firecracker pid, and open an ssh connection.
+    firecracker_pid = test_microvm.jailer_clone_pid
+    ssh_connection = net_tools.SSHConnection(test_microvm.ssh_config)
+
+    # Get initial rss consumption.
+    init_rss = get_rss_mem_by_pid(firecracker_pid)
+
+    # Dirty memory, then inflate balloon and get ballooned rss consumption.
+    make_guest_dirty_memory(ssh_connection)
+
+    response = test_microvm.balloon.patch(
+        num_pages=35840,
+        stats_polling_interval_s=0
+    )
+
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    time.sleep(5)
+    balloon_rss = get_rss_mem_by_pid(firecracker_pid)
+
+    # Check that the ballooning reclaimed the memory.
+    assert balloon_rss - init_rss <= 10000
+
+
+# pylint: disable=C0103
+def test_inflate_reduces_free(test_microvm_with_ssh_and_balloon,
+                              network_config):
+    """Verify that the balloon statistics work as expected."""
+    test_microvm = test_microvm_with_ssh_and_balloon
+    test_microvm.spawn()
+    test_microvm.basic_config()
+    _tap, _, _ = test_microvm.ssh_network_config(network_config, '1')
+    test_microvm.ssh_config['ssh_key_path'] = os.path.join(
+        test_microvm.fsfiles,
+        'debian.rootfs.id_rsa'
+    )
+
+    # Install deflated balloon.
+    response = test_microvm.balloon.put(
+        num_pages=0,
+        deflate_on_oom=False,
+        must_tell_host=False,
+        stats_polling_interval_s=1
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    # Start the microvm
+    test_microvm.start()
+
+    # Get and open an ssh connection.
+    ssh_connection = net_tools.SSHConnection(test_microvm.ssh_config)
+
+    # Get the free memory before ballooning.
+    available_mem_deflated = get_free_mem_ssh(ssh_connection)
+
+    # Inflate 64 MB == 16384 page balloon.
+    response = test_microvm.balloon.patch(
+        num_pages=16384,
+        stats_polling_interval_s=0
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+    time.sleep(5)
```

Comment:
Do we really need to sleep 5 seconds and stall the CI pipeline ?

Reply:
We do because the time it takes the guest driver to actually fulfill the device's requests is very volatile. I have tried multiple values for this and even implemented a system which periodically checks whether the assert is successful if it wasn't in the past, but this was actually the best option out of what I've tried.

Reply:
Sleep is not a good way to synchronize expectations with the guest. I am concerned about intermittent failures.

Reply:
There is no way you can get notified by the firecracker process that the balloon driver has done the work. We can only poll the RSS and watch for changes. This chosen waiting time (5 seconds) is what I have (subjectively) determined to be the time I am willing to let the driver do something. If there is no result after the waiting time has passed, the test will consider it a failure.

The only other option that I see available is polling every second until we get the assert right, but I think it's messier.